pyhap/accessory_driver.py

```diff
@@ -859,118 +860,99 @@ def set_characteristics(self, chars_query, client_addr):
-                    continue
-                char_to_iid[char] = iid
-                updates_by_service.setdefault(char.service, {}).update({char: value})
+        callback_results = {}
+
+        # Proccess accessory and service level setter callbacks
+        for acc, updates_by_service in updates_by_accessories_services.items():
+            aid = acc.aid
 
             # Accessory level setter callbacks
+            acc_set_result = None
             if acc.setter_callback:
-                set_result = _wrap_acc_setter(acc, updates_by_service, client_addr)
-                if set_result != HAP_SERVER_STATUS.SUCCESS:
-                    had_error = True
-                for iid in updates[aid]:
-                    setter_results[aid][iid] = set_result
+                acc_set_result = _wrap_acc_setter(acc, updates_by_service, client_addr)
 
             # Service level setter callbacks
             for service, chars in updates_by_service.items():
-                if not service.setter_callback:
-                    continue
-                set_result = _wrap_service_setter(service, chars, client_addr)
-                if set_result != HAP_SERVER_STATUS.SUCCESS:
-                    had_error = True
-                for char in chars:
-                    setter_results[aid][char_to_iid[char]] = set_result
-
-        if not had_error and not had_write_response:
-            return None
-
-        return {
-            HAP_REPR_CHARS: [
-                {
-                    HAP_REPR_AID: aid,
-                    HAP_REPR_IID: iid,
-                    HAP_REPR_STATUS: status,
-                    **(
-                        {HAP_REPR_VALUE: setter_responses[aid][iid]}
-                        if setter_responses.get(aid, {}).get(iid, None) is not None
-                        else {}
-                    )
+                char_set_result = None
+                if service.setter_callback:
+                    char_set_result = _wrap_service_setter(service, chars, client_addr)
+                set_result = char_set_result or acc_set_result
+
+                callback_results[aid] = {
+                    char_to_iid[char]: {HAP_REPR_STATUS: set_result}
+                    for char in chars if set_result is not None
                 }
-                for aid, iid_status in setter_results.items()
-                for iid, status in iid_status.items()
-            ]
-        }
+
+        results = [
+            {
+                HAP_REPR_AID: aid,
+                HAP_REPR_IID: iid,
+                **query_results,
+                **callback_results.get(aid, {}).get(iid, {})
+            }
+            for aid, query_iid_results in query_results.items()
+            for iid, query_results in query_iid_results.items()
+        ]
+        nonempty_results_exist = any(
+            result[HAP_REPR_STATUS] != HAP_SERVER_STATUS.SUCCESS
+            or HAP_REPR_VALUE in result
+            for result in results
+        )
```

Comment:
Its probably a bit more readable (and faster) to unpack this into a single loop since the generator expression is having to iterate everything again

Reply:
Ran a benchmark via pytest-benchmark. On low input sizes (<10) list-append-based solution indeed has a ~1us mean gain (M1 Pro) over this one. 

Didn't consider that queries are not that big usually, so append solution does not cause a resize. Gonna change.